plugins/repository-hdfs/src/main/java/org/elasticsearch/plugin/hadoop/hdfs/Utils.java

```diff
@@ -0,0 +1,84 @@
+package org.elasticsearch.plugin.hadoop.hdfs;
+
+import java.net.URL;
+import java.security.AccessControlContext;
+import java.security.AccessController;
+import java.security.DomainCombiner;
+import java.security.PrivilegedAction;
+import java.security.ProtectionDomain;
+
+import org.elasticsearch.SpecialPermission;
+
+public abstract class Utils {
```

Comment:
The reason for this class existing is so it makes the AccessControlContext of the plugin (based on its codeSource) available to the nested Hadoop classloader (which sits in a different ClassLoader and different codeSource).
The classes need to communicate with each other across ClassLoaders so I went the route of security by obscurity which is, use this class that offers the ACC through a protected method so that the a class in the nested ClassLoader can extend it and thus gain access to its method and ACC context.
In the end this means gaining access to the combiner which checks anyway for the codeSource of the assigned ProtectionDomain so even if other classes would gain access to that, it wouldn't really serve them.

The other alternative was to use split packages - have a class in the same package as HdfsPlugin, but loaded inside the nested Hadoop CL. This way it would not have the privileges of the plugin but it could access to its method (which would be package private instead of protected) and thus gain access to it.
It's more obscure/secure but to the point where folks looking at it might wonder what is going on...